test/jdk/java/net/httpclient/http2/IdleConnectionTimeoutTest.java

```diff
@@ -154,40 +162,72 @@ private HttpResponse<String> runRequest(HttpClient hc, HttpRequest req, int slee
 
     static class ServerTimeoutHandler implements Http2Handler {
 
-        InetSocketAddress remote;
+        boolean acceptedFirstConnection;
+        int firstConnectionHash = 0;
 
         @Override
         public void handle(Http2TestExchange exchange) throws IOException {
-            if (remote == null) {
-                remote = exchange.getRemoteAddress();
-                exchange.sendResponseHeaders(200, 0);
-            } else if (!remote.equals(exchange.getRemoteAddress())) {
-                testLog.println("ServerTimeoutHandler: New Connection was used, idleConnectionTimeoutEvent fired."
-                        + " First remote: " + remote + ", Second Remote: " + exchange.getRemoteAddress());
-                exchange.sendResponseHeaders(200, 0);
-            } else {
-                exchange.sendResponseHeaders(400, 0);
+            if (exchange instanceof TestExchangeSupplier exch) {
+                if (!acceptedFirstConnection) {
+                    firstConnectionHash = exch.getTestConnection().hashCode();
+                    acceptedFirstConnection = true;
+                    exch.sendResponseHeaders(200, 0);
+                } else {
+                    int secondConnectionHash = exch.getTestConnection().hashCode();
+                    int cmp = Integer.compare(firstConnectionHash, secondConnectionHash);
+
+                    if (cmp < 0 | cmp > 0) {
```

Comment:
In other words, `if (cmp != 0)` ? ;-)

It may be simpler and less error prone to save the connection (as an Object) in the handler.
```
        volatile Object previousConnection = null;
```

Then the condition become:
```
                if (previousConnection == null) {
                    previousConnection = exch.getTestConnection();
                    exch.sendResponseHeaders(200, 0);
                } else {
                    var secondConnection = exch.getTestConnection();

                    if (previousConnection != secondConnection) {
                        ....
```

Reply:
I could be wrong but I think I might have tried that before what you're calling `previousConnection` in the sample code you've given gets set to `null` in the time between the first and second requests during the tests where a timeout fires. Which does serve to test that the connections are different to be fair but maybe for some reason `previousConnection` could be null that is unrelated to it being closed by the test server. For example `createConnection()` might close early due to a Reset Stream or Protocol Error. 

I know that's very edge case but its the reason I compared using the hash code. I'm open to changing it though of course! Might be less error prone as you say.

Also, very fair on the `if (cmp != 0)` suggestion 😅 I'll definitely shorten it to that if I stick with the comparison as it is right now.

Reply:
I don't see how using the hash wouldn't cause the exact same issue. Also you really need volatile here, since two different requests are going to be handled by two different threads.